### ISSUE TITLE
MAML().adapt() cause a gradual decline in training speed

### ISSUE THREAD

**Issue by lzl19971215** [learnables/learn2learn#230]
Hi,
I use `MAML.adapt` to update my cloned learner with first order approximation, and find the training speed decreases with the number of training steps  in each iteration. 
My code is as follow:
```
maml = l2l.algorithms.MAML(model, lr=fast_lr, first_order=True)
learner = maml.clone()

# in each iteration:
for batch in loader:
  ...
  train_error = criterion(pred, label)
  learner.adapt(train_error)
```

**Comment by seba-1511** [comment]
Hello @lzl19971215 ,

Could you specify how you measure the decline in training speed?

If you mean that adaptation step `n` is slower than adaptation step `n-1`, that's' a known bug with PyTorch, see these issues:

* https://github.com/pytorch/pytorch/issues/12635
* https://github.com/pytorch/pytorch/pull/52180
* https://github.com/facebookresearch/higher/issues/20

**Comment by lzl19971215** [comment]
thanks for your reply!
But I don't think its this bug, for i compare using `maml.adapt`  with not using maml to wrap my model but `train_error.backward()` and 'optimizer.step()' to update. Both of them adapt the model on 5 dataloaders for 2 adaptation steps.

Using `MAML.adapt`
![adapt](https://user-images.githubusercontent.com/63151530/109603044-374e0600-7b5c-11eb-91f1-38950a52e354.jpg)
I found  training on same loader in adaptation step 2 was slower than that in step 1.

Using 'error.backward'
![backward](https://user-images.githubusercontent.com/63151530/109603047-3a48f680-7b5c-11eb-95dd-2996c18cb28c.jpg)
This time training on same loader in two steps dont have much different speed.
![backward](https://user-images.githubusercontent.com/63151530/109604051-d1fb1480-7b5d-11eb-96c9-3277f57fc332.jpg)

**Comment by seba-1511** [comment]
If I understand correctly, this is indeed the same bug. If you replace `learner.adapt` by `error.backward` the model is not updated MAML-style (and does not track gradients of the adaptation steps) and so you're back to standard training -- i.e. no slow-down.

**Comment by lzl19971215** [comment]
I think setting `first_order` to True will omit second derivaties, and so in this case, do we still need to track adaptation steps gradients?

**Comment by seba-1511** [comment]
Correct, if you set `first_order=True` this is equivalent to not tracking the gradient of adaptation steps.
